dojo/finding_group/redis.py

```diff
@@ -0,0 +1,256 @@
+
+    @staticmethod
+    def add_finding(finding: Finding, mode: GroupMode) -> None:
+        finding_groups = DynamicFindingGroups.get_group_names(finding, mode)
+        if not finding_groups:
+            return
+        redis_client = get_redis_client()
+        for finding_group_name in finding_groups:
+            finding_group_id = base64.b64encode(finding_group_name.encode()).decode()
+            fg_key = DynamicFindingGroups.get_fg_key(mode)
+            id_map_key = DynamicFindingGroups.get_id_map_key(mode)
+
+            finding_group = DynamicFindingGroups.load_from_id(finding_group_id, fg_key)
+            if not finding_group:
+                finding_group = DynamicFindingGroups(
+                    finding_group_id=finding_group_id,
+                    name=finding_group_name,
+                )
+
+            if finding.id not in finding_group.finding_ids:
+                finding_group.add(finding)
+
+            redis_client.hset(fg_key, finding_group_id, json.dumps(finding_group.to_dict()))
+            group_ids_raw = redis_client.hget(id_map_key, finding.id)
+            group_ids = json.loads(group_ids_raw) if group_ids_raw else []
+            if finding_group_id not in group_ids:
+                group_ids.append(finding_group_id)
+            redis_client.hset(id_map_key, finding.id, json.dumps(group_ids))
+
+    # This method is used in finding_groups table to show SLA
+    def get_days_remaining(self) -> str:
+        if self.sla_finding_id:
+            finding = Finding.objects.filter(id=self.sla_finding_id).first()
+            days_remaining = finding.sla_days_remaining()
+            severity = finding.severity
+            sla_start_date = finding.get_sla_start_date().strftime("%b %d, %Y")
+            status = "age-green"
+            status_text = f"Remediation for {severity.lower()} findings due in {days_remaining} days or less (started {sla_start_date})"
+            if days_remaining and days_remaining < 0:
+                status = "age-red"
+                status_text = f"Overdue: Remediation for {severity.lower()} findings overdue {days_remaining} days (started {sla_start_date})"
+                days_remaining = abs(days_remaining)
+        elif any(
+            Finding.objects.filter(
+                id__in=self.finding_ids,
+                active=True,
+            ),
+        ):
+            status = "severity-Info"
+            status_text = "No SLA set, but at least one finding is active"
+            days_remaining = "No SLA"
+        else:
+            status = "age-blue"
+            status_text = "No active finding"
+            days_remaining = "Concluded"
+        title = (
+            f'<a class="has-popover" data-toggle="tooltip" data-placement="bottom" title="" href="#" data-content="{status_text}">'
+            f'<span class="label severity {status}">{days_remaining}</span></a>'
+        )
+        return mark_safe(title)
```

Comment:
I wonder if this could be made more clean by maybe putting it in a display tag to avoid mixing logic and display too much? I also wonder about the overlap there is with the Finding_Group model which also has logic to determine SLA data (and severity?) for a list of findings.